pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -846,61 +847,169 @@ func (m *kubeGenericRuntimeManager) purgeInitContainers(ctx context.Context, pod
+		status := podStatus.FindContainerStatusByName(container.Name)
+		if status == nil {
+			continue
+		}
+		switch status.State {
+		case kubecontainer.ContainerStateCreated,
+			kubecontainer.ContainerStateRunning,
+			kubecontainer.ContainerStateExited:
+			return true
+		default:
+			// Ignore other states
+		}
+	}
+	return false
+}
+
+// computeInitContainerActions sets the actions on the given changes that need
+// to be taken for the init containers. This includes actions to initialize the
+// init containers and actions to keep sidecar containers running.
+// computeInitContainerActions returns true if pod has been initialized.
+//
+// The actions include:
+// - Start the first init container that has not been started.
+// - Restart all sidecar containers that have started but are not running.
+// - Kill the sidecar containers that have failed the startup probe.
+func (m *kubeGenericRuntimeManager) computeInitContainerActions(pod *v1.Pod, podStatus *kubecontainer.PodStatus, changes *podActions) bool {
 	if len(pod.Spec.InitContainers) == 0 {
-		return nil, nil, true
+		return true
 	}
 
 	// If any of the main containers have status and are Running, then all init containers must
 	// have been executed at some point in the past.  However, they could have been removed
 	// from the container runtime now, and if we proceed, it would appear as if they
-	// never ran and will re-execute improperly.
-	for i := range pod.Spec.Containers {
-		container := &pod.Spec.Containers[i]
-		status := podStatus.FindContainerStatusByName(container.Name)
-		if status != nil && status.State == kubecontainer.ContainerStateRunning {
-			return nil, nil, true
-		}
-	}
+	// never ran and will re-execute improperly except for the sidecar containers.
+	podHasInitialized := hasAnyRegularContainerCreated(pod, podStatus)
 
-	// If there are failed containers, return the status of the last failed one.
-	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
+	previousContainerInitialized := true
+	restartOnFailure := shouldRestartOnFailure(pod)
+	for i := range pod.Spec.InitContainers {
 		container := &pod.Spec.InitContainers[i]
-		status := podStatus.FindContainerStatusByName(container.Name)
-		if status != nil && isInitContainerFailed(status) {
-			return status, container, false
+		if podHasInitialized && !types.IsSidecarContainer(container) {
+			// after initialization, only sidecar containers need to be kept running
+			continue
 		}
-	}
 
-	// There are no failed containers now.
-	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
-		container := &pod.Spec.InitContainers[i]
 		status := podStatus.FindContainerStatusByName(container.Name)
+		klog.V(4).InfoS("Computing init container action", "pod", klog.KObj(pod), "container", container.Name, "status", status)
+		// If the container is not found, it means it has not been started yet.
 		if status == nil {
-			continue
+			if podHasInitialized || previousContainerInitialized {
```

Comment:
At least at first glance, this logic seems wrong (but I’m possibly just missing the new logic).

Before, we only needed to find the latest initialized container (we walked backwards from the end of the list to the first one with status) - we don’t need containers 0-N to have history, only N+1.  Now we stop at the first container with missing status?

Described another way, a pod with N init containers requires only O(1) init container statuses to be preserved (the last one) before.  After the changes, we need to preserve all N?  That would not be desirable if true.

Reply:
Are you concerned more about after a Pod is fully initialized? Or during initialization as well? After fully initialized we can check the status of a first regular container (see `podHasInitialized` above).

We will need to keep status of all sidecars because they become a regular containers after initialization. But we can definitely optimize for the Init containers the same way as before for the initialization stage. Do you think it is critical?

Reply:
I think this change is needed as the sidecar container should behave like a regular container once it's started.

let's say there is a pod with,
```
initContainers:
- sidecar-0: has started but terminated
- init-container-1: initialized
- init-container-2: not started yes
```
we should RESTART sidecar-0 and initialize init-container-2.

So we need to start iterating from 0-index.

Reply:
> We will need to keep status of all sidecars because they become a regular containers after initialization. But we can definitely optimize for the Init containers the same way as before for the initialization stage. Do you think it is critical?

we can optimize it in that way, but is it needed? I think it ends up iterating the initContainers twice..?

Reply:
> > We will need to keep status of all sidecars because they become a regular containers after initialization. But we can definitely optimize for the Init containers the same way as before for the initialization stage. Do you think it is critical?
> 
> we can optimize it in that way, but is it needed? I think it ends up iterating the initContainers twice..?

Not necessarily. We start from the end. Once we found Init container that is already started, start one after it. But keep iterating to find all sidecar containers that may need to be restarted

Reply:
> I think the current way is slightly better in terms of the number of iterations. What do you think?
> current: O(initialized containers)
> proposed: O(all init containers)

I think the @smarterclayton's concern is that if Init containers are very long running or we have a lot of them, after the kubelet restart, the status of first containers may get lost, while the status of last executing one will be preserved by runtime. So we need to look up from the end.

Reply:
> I think the @smarterclayton's concern is that if Init containers are very long running or we have a lot of them, after the kubelet restart, the status of first containers may get lost, while the status of last executing one will be preserved by runtime. So we need to look up from the end.

I agree that that way is safer, as at least it doesn't change the existing behavior. Let's take that approach.

---

Anyway, I don't know if the other part of the kubelet can handle that case. If there is any init container that does not have a state, the kubelet will consider it a non-running container and not restart it, and the pod can be stuck in the `Pending` phase forever.
https://github.com/kubernetes/kubernetes/blob/988094878e0a5c8e9de8d1fc22bae2a2236628d7/pkg/kubelet/kubelet_pods.go#L1784-L1792

Reply:
Sorry for not giving more explanation of why this was important:

1. keeping exited containers around puts load on the node - the filesystem, tracking data in CRI, and in the kubelet any O(pods) loops
2. because pods are general purpose, people can create pods with lots and lots of init containers, and they can use them in restartnever aka jobs - e.g. modelling a build process as a series of init containers and the regular container being the one that completes the process, which might have 10+ init containers and there might be hundreds or thousands of these type of pods per node (because the system is running lots of small builds) - that consumes lots of resources
3. we *try* to keep containers around after pods complete so users can view logs / potentially access filesystem contents (i.e. system log agents / tooling that's acting on disk) until they aren't needed
4. the kube garbage collection process periodically purges those containers as resources are needed, so occasionally on a busy node you'll have a scenario where kubelet needs to delete completed containers *before* the pod is completed (say if the last init container step takes a long time)
5. the kubelet doesn't have a checkpointing mechanism - so our "state" is the previously completed containers for tracking how far we've gotten on initialization when kubelet is restarted
6. therefore we only want to write logic for init containers that relies on the minimal number of init containers - the last one - and we do that with the reverse loop

We should really add a comment in this method and right above the loop (In the new code) that makes it clear why the reverse iteration is so important.

In regards to your last question, we should definitely make sure this isn't happening, but the way the logic for that method should work is 

1. once a pod is running regular containers, we assume that initialization was succeeded and ignore those statuses
2. if a pod has no running containers, then only consider the last one

The method you describe is only responsible for merging the previously recorded state of each container status from the API to our latest status (from the container runtime).   The method that determines what phase the pod is in is [Kubelet.generateAPIPodStatus](https://github.com/kubernetes/kubernetes/blob/2cc529282a37e5a8aa6479ce71f8660645b18f52/pkg/kubelet/kubelet_pods.go#L1534) which calls `getPhase` which [checks whether init containers are being initalized, but only checks waiting status for regular containers](https://github.com/kubernetes/kubernetes/blob/2cc529282a37e5a8aa6479ce71f8660645b18f52/pkg/kubelet/kubelet_pods.go#L1397).  I.e. if all regular containers are running, the pod is running, otherwise the pod is pending.

So the scenario you describe shouldn't happen, because after the kubelet is restarted we should be looking at whatever the last container is and only starting ones after that, and if we are running *any* init containers there should be no running containers.  If after a restart, we're only running *some* of the running containers the SyncPod method should simply start the rest of them, and in *that* scenario we will briefly go back to pending (which is the correct behavior, pods are allowed to go Running -> Pending for lots of reasons including loss of status, restarts, cleanup of node state during reboot).  

It's important to remember that kubelets are considered "stateless" and pods can be executed multiple times without updating the api because of that - this was deliberate because users who need "at most once" execution can implement the logic themselves inside their pod, and the vast majority of workloads do not need that very strong logic.

Reply:
Thanks for the explanation.

> The method that determines what phase the pod is in is [Kubelet.generateAPIPodStatus](https://github.com/kubernetes/kubernetes/blob/2cc529282a37e5a8aa6479ce71f8660645b18f52/pkg/kubelet/kubelet_pods.go#L1534) which calls `getPhase` which [checks whether init containers are being initalized, but only checks waiting status for regular containers](https://github.com/kubernetes/kubernetes/blob/2cc529282a37e5a8aa6479ce71f8660645b18f52/pkg/kubelet/kubelet_pods.go#L1397). I.e. if all regular containers are running, the pod is running, otherwise the pod is pending.


I think the current implementation of `getPhase` [takes the `pendingInitialization` into account which is increased when there is any pending init container.](https://github.com/kubernetes/kubernetes/blob/2cc529282a37e5a8aa6479ce71f8660645b18f52/pkg/kubelet/kubelet_pods.go#L1459-L1465) So, if there is any init container with `Waiting` state (which is a default state when there is no container status from the container runtime), the phase will be `Pending`.

I agree that it is important to make kubelet work with minimum information, but I think the current kubelet cannot handle those cases correctly. I think we may fix it separately from this PR.

Reply:
> > The method that determines what phase the pod is in is [Kubelet.generateAPIPodStatus](https://github.com/kubernetes/kubernetes/blob/2cc529282a37e5a8aa6479ce71f8660645b18f52/pkg/kubelet/kubelet_pods.go#L1534) which calls `getPhase` which [checks whether init containers are being initalized, but only checks waiting status for regular containers](https://github.com/kubernetes/kubernetes/blob/2cc529282a37e5a8aa6479ce71f8660645b18f52/pkg/kubelet/kubelet_pods.go#L1397). I.e. if all regular containers are running, the pod is running, otherwise the pod is pending.
> 
> I think the current implementation of `getPhase` [takes the `pendingInitialization` into account which is increased when there is any pending init container.](https://github.com/kubernetes/kubernetes/blob/2cc529282a37e5a8aa6479ce71f8660645b18f52/pkg/kubelet/kubelet_pods.go#L1459-L1465) So, if there is any init container with `Waiting` state (which is a default state when there is no container status from the container runtime), the phase will be `Pending`.
> 
> I agree that it is important to make kubelet work with minimum information, but I think the current kubelet cannot handle those cases correctly. I think we may fix it separately from this PR.

Sorry for the confusion from the previous comment (I'll hide it). I tested the scenario locally and found that there is no problem to fix.

```yaml
initContainers:
- name: init-1
- name: init-2
containers:
- name: regular
```

1. create a pod with the above spec.
2. wait for the pod to be initialized
3. delete init-1 init container state with `cricrl rm`

found that the kubelet stops getting the status of the init-1 from the container runtime but it doesn't matter, because if the previous status is `Terminated` and there is no state from the container runtime, the kubelet doesn't update those container's status anymore.

It will work fine as you both said. I'll just update this code to traverse in reverse.

Reply:
fixed it to iterate through in rever order